File: 2022/july/04-07-2022/candy.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        left = [1] * n
        right = [1] * n

        for i in range(1, n):
            if ratings[i] > ratings[i - 1]:
                left[i] = left[i - 1] + 1

        for j in range(n - 2, -1, -1):
            if ratings[j] > ratings[j + 1]:
                right[j] = right[j + 1] + 1

        minimumCandies = 0

        for i in range(n):
            minimumCandies += max(left[i], right[i])

        return minimumCandies

File: 2022/july/04-07-2022/candy.rs
impl Solution {
    pub fn candy(ratings: Vec<i32>) -> i32 {
        let N = ratings.len();
        let mut candies = vec![ 1 as i32; N ];
        // forward pass
        for i in 1 .. N {
            if ratings[i] > ratings[i-1] {
                candies[i] = candies[i-1] + 1;
            }
        }
        // backwards pass
        for i in (0 .. N-1).rev() {
            if ratings[i] > ratings[i+1] {
                candies[i] = candies[i].max(candies[i+1] + 1);
            }
        }
        candies.iter().sum()
    }
}